pkgs/stdenv/generic/check-meta.nix

```diff
@@ -247,16 +247,7 @@ let
       isEnabled = lib.findFirst (x: x == reason) null showWarnings;
     in if isEnabled != null then builtins.trace msg true else true;
 
-
-  # A shallow type check. We are using NixOS'
-  # option types here, which however have the major drawback
-  # of not providing full type checking (part of the type check is
-  # done by the module evaluation itself). Therefore, the checks
-  # will not recurse into attributes.
-  # We still provide the full type for documentation
-  # purposes and in the hope that they will be used eventually.
-  # See https://github.com/NixOS/nixpkgs/pull/191171 for an attempt
-  # to fix this, or mkOptionType in lib/types.nix for more information.
+  # TODO make this into a proper module and use the generic option documentation generation?
```

Comment:
Wouldn't disallowing raw module syntax be as easy as `{ ... }: { config = meta; }` ?

Somewhat tangentially, the module system is already leaking in an undesirable way:

```
nix-repl> lib.typeCheck lib.types.str (lib.mkIf true "a")
true
```

This is hard to fix because `type.check` and `mergeDefinitions` are so tightly coupled.

Reply:
You're right. We should be checking type outputs, not type inputs, and there's not type method for that yet.